### PR TITLE
Add ML option for coarse solver

### PR DESCRIPTION
**Pull request by aprokop**
It's running ML now, I can see the output. Next steps:
- [x] propagate parameters from `boost::property_tree` to `Teuchos::ParameterList`
  Control the coarse levels from input deck. Hopefully not too hard, as ML parameters are flat.
- [x] Fix operator and grid complexities
  If one uses AMG for coarse levels too, it needs to combine its complexity with amge
- [x] Double check whether we solve the right thing
  ~Should ML be used as a preconditioner, or as a standalone solver? Do we pass residual or rhs?~ ML can be use as either a preconditioner, or as a standalone with a single iteration. There is no need to compute residual, it can be thought of having a zero initial guess.
- [ ] Study the convergence

### CODE REVIEW

**Comment by aprokop** [comment]
It's running ML now, I can see the output. Next steps:
- [ ] propagate parameters from `boost::property_tree` to `Teuchos::ParameterList`
  Control the coarse levels from input deck. Hopefully not too hard, as ML parameters are flat.
- [ ] Fix operator and grid complexities
  If one uses AMG for coarse levels too, it needs to combine its complexity with amge
- [ ] Double check whether we solve the right thing
  Should ML be used as a preconditioner, or as a standalone solver? Do we pass residual or rhs?
- [ ] Study the convergence

**Comment by dalg24** [comment]
Try something like that
```C++
boost::property_tree::ptree ptree; // assume it contains the data structure
std::stringstream ss;
boost::property_tree::xml_parser::write_xml( ss, ptree );
auto parameter_list = 
// NOTE I couldn't find a static function to parse the stream in Teuchos :/
Teuchos::ParameterList dummy;
auto plist = dummy.getParametersFromXmlString( ss.str() );
```

**Comment by aprokop** [comment]
I think this is ready for review. It runs and converges, but I have not designed the test yet as I need to study the convergence.

File: source/utils.cc [hunk 1]
```diff
@@ -41,4 +42,79 @@ void matrix_market_output_file(
...
+
+static bool is_integer(const std::string &s)
+{
+  std::istringstream is(s);
+  int i;
+  is >> i;
+  return !is.fail() && is.eof();
+}
+
+static bool is_double(const std::string &s)
```
Comment (dalg24) [inline source/utils.cc]: I think `(boost::property_tree::ptree::get_value_optional<T>()` does what you wanted.

File: tests/test_utils.cc [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+#include <Teuchos_ParameterList.hpp>
+#include <boost/property_tree/info_parser.hpp>
+#include <boost/property_tree/ptree.hpp>
+
+BOOST_AUTO_TEST_CASE(plist2ptree)
+{
+  boost::property_tree::ptree ptree;
+  Teuchos::ParameterList plist;
+
+  std::string info_string = " \n\
```
Comment (dalg24) [inline tests/test_utils.cc]: You might want to consider using a raw string literal here:
```C++
std::string info_string = R"(
eigensolver
{
  "number of eigenvectors" 2
  tolerance 1e-14
}
)"
```

**Comment by Rombur** [comment]
Can you add a test that uses ML.

**Comment by aprokop** [comment]
@Rombur Of course, but I need to check few things first.

**Comment by aprokop** [comment]
@dalg24 Thank you for the comments, that worked!

**Comment by aprokop** [comment]
@Rombur Added ML test. Ready for final review.

File: include/mfmg/dealii_operator.hpp [hunk 3]
```diff
@@ -122,9 +126,14 @@ class DealIISmootherOperator : public Operator<VectorType>
                          std::shared_ptr<boost::property_tree::ptree> params);
 
   virtual size_t m() const override final { return _matrix.m(); }
-
   virtual size_t n() const override final { return _matrix.n(); }
 
+  virtual size_t grid_complexity() const override final { return m(); }
+  virtual size_t operator_complexity() const override final
+  {
+    ASSERT_THROW_NOT_IMPLEMENTED();
```
Comment (Rombur) [inline include/mfmg/dealii_operator.hpp]: Need a return to avoid warning

Comment (aprokop) [reply include/mfmg/dealii_operator.hpp]: Damn, it was there! I guess I messed up when cleaning and rebasing.

**Pull request merged**